community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsTracker.java

```diff
@@ -300,56 +449,163 @@ public void close()
-                // select the next file, and create a writer for it
-                try ( CountsStore.Writer<CountsKey,CopyableDoubleLongRegister> writer =
-                              nextWriter( state, lastCommittedTxId ) )
-                {
-                    state.accept( writer );
-                    // replaceSecond the old store with the
-                    this.state = new ConcurrentCountsTrackerState( writer.openForReading() );
-                }
-                logger.debug( "Completed writing of counts store with txId=" + lastCommittedTxId );
-                // close the old store
-                state.close();
+                this.txId = txId;
             }
         }
     }
 
-    CountsStore.Writer<CountsKey,CopyableDoubleLongRegister> nextWriter( CountsTrackerState state, long lastTxId )
-            throws IOException
+    private static final class TxId
+    {
+        private final long txId;
+        private final long minorVersion;
+
+        TxId( long txId, long minorVersion )
+        {
+
+            this.txId = txId;
+            this.minorVersion = minorVersion;
+        }
+    }
+
+    private static class UpdateFirstValue extends Update<CountsKey>
+    {
+        private static final int OFFSET = 0;
+        private final long delta;
+
+        UpdateFirstValue( CountsKey key, long delta )
+        {
+            super( key );
+            this.delta = delta;
+        }
+
+        @Override
+        protected void update( WritableBuffer value )
+        {
+            value.putLong( OFFSET, value.getLong( OFFSET ) + delta );
+        }
+    }
+
+    private static class UpdateSecondValue extends Update<CountsKey>
     {
-        if ( alphaFile.equals( state.storeFile() ) )
+        private static final int OFFSET = 8;
+        private final long delta;
+
+        UpdateSecondValue( CountsKey key, long delta )
         {
-            return state.newWriter( betaFile, lastTxId );
+            super( key );
+            this.delta = delta;
         }
-        else
+
+        @Override
+        protected void update( WritableBuffer value )
         {
-            return state.newWriter( alphaFile, lastTxId );
+            value.putLong( OFFSET, value.getLong( OFFSET ) + delta );
         }
     }
 
-    File storeFile()
+    private static class AssignValues extends Update<CountsKey>
     {
-        return state.storeFile();
+        private final long first, second;
+
+        public AssignValues( CountsKey key, long first, long second )
+        {
+            super( key );
+            this.first = first;
+            this.second = second;
+        }
+
+        @Override
+        protected void update( WritableBuffer value )
+        {
+            value.putLong( 0, first );
+            value.putLong( 8, second );
+        }
     }
 
-    private static File storeFile( File base, String version )
+    private class ValueRegister extends Reader<Register.DoubleLongRegister>
```

Comment:
There are way too many private static classes in here, maybe time to create a sub-package for all this stuff?